Cython/Utils.py

```diff
@@ -316,7 +330,14 @@ def find_versioned_file(directory, filename, suffix,
     assert not suffix or suffix[:1] == '.'
     path_prefix = os.path.join(directory, filename)
 
-    matching_files = glob.glob(path_prefix + ".cython-*" + suffix)
+    if suffix == '.pxd':
+        matching_files = glob.glob(path_prefix + ".cython-*" + suffix)
+        _parse_file_version = _pxd_parse_file_version
+    elif suffix == '.so' or suffix == '.pyd':
+        matching_files = glob.glob(path_prefix + ".*-*" + suffix)
+        _parse_file_version = _so_parse_file_version
+    else:
+        matching_files = []
```

Comment:
Is there a reason why we need to change this? This was meant purely to find versioned .pxd files on cimports. Wasn't it enough to extend the list of package filenames?

I also don't see a test for this.

Reply:
> Is there a reason why we need to change this?

Yes it is used in `self.search_include_directories()` called here:

https://github.com/cython/cython/pull/5707/files#diff-76076a7a593cd1c5b170b7bf9a6c215bf2b8e39733b0d08e778b78e7cc004676R253

>  This was meant purely to find versioned .pxd files on cimports. Wasn't it enough to extend the list of package filenames?

No, the `self.search_include_directories()` is used to find `__init__.XYZ` file using this function. Hence we need to extend the function to support also `.so` files.

> I also don't see a test for this.

test for `.so` file is attached to this PR. Test for `.py` file is already present in test suit.

Reply:
But we only searched for `.py` files there, apparently, not even for `.pxd`. That would mean that this never worked for `__init__.pxd` files, but I'm pretty sure that we had this working for a while, so I'm very surprised that this needs changing.

I would prefer not mixing different versioning schemes in this function. It was meant for Cython versions, not CPython/PyPy/whatever versions, and it doesn't feel right to extend it to that.